src/gen/run_block_generator.rs

```diff
@@ -75,3 +78,111 @@ pub fn run_block_generator<GenBuf: AsRef<[u8]>>(
+    a: &Allocator,
+    mut n: NodePtr,
+    e: ErrorCode,
+) -> Result<[NodePtr; N], ValidationErr> {
+    let mut ret: [NodePtr; N] = [0; N];
+    let mut counter = 0;
+    assert!(N > 0);
+    while let Some((item, rest)) = a.next(n) {
+        if counter == N - 1 {
+            break;
+        }
+        n = rest;
+        ret[counter] = item;
+        counter += 1;
+    }
+    if counter != N - 1 {
+        return Err(ValidationErr(n, e));
+    }
+    ret[counter] = n;
+    Ok(ret)
+}
+
+// This has the same behavior as run_block_generator() but implements the
+// generator ROM in rust instead of using the CLVM implementation.
+// it is not backwards compatible in the CLVM cost computation (in this version
+// you only pay cost for the generator, the puzzles and the conditions).
+// it also does not apply the stack depth or object allocation limits the same,
+// as each puzzle run in its own environment.
+pub fn run_block_generator2<GenBuf: AsRef<[u8]>>(
+    a: &mut Allocator,
+    program: &[u8],
+    block_refs: &[GenBuf],
+    max_cost: u64,
+    flags: u32,
+) -> Result<SpendBundleConditions, ValidationErr> {
+    let byte_cost = program.len() as u64 * COST_PER_BYTE;
+
+    let mut cost_left = max_cost;
+    subtract_cost(a, &mut cost_left, byte_cost)?;
+
+    let clvm_deserializer = node_from_bytes(a, &CLVM_DESERIALIZER)?;
+    let program = if (flags & ALLOW_BACKREFS) != 0 {
+        node_from_bytes_backrefs(a, program)?
+    } else {
+        node_from_bytes(a, program)?
+    };
+
+    // iterate in reverse order since we're building a linked list from
+    // the tail
+    let mut blocks = a.null();
+    for g in block_refs.iter().rev() {
+        let ref_gen = a.new_atom(g.as_ref())?;
+        blocks = a.new_pair(ref_gen, blocks)?;
+    }
+
+    // the first argument to the generator is the serializer, followed by a list
+    // of the blocks it requested.
+    let mut args = a.new_pair(blocks, a.null())?;
+    args = a.new_pair(clvm_deserializer, args)?;
+
+    let dialect = ChiaDialect::new(flags);
+
+    let Reduction(clvm_cost, mut all_spends) = run_program(a, &dialect, program, args, cost_left)?;
+
+    subtract_cost(a, &mut cost_left, clvm_cost)?;
+    all_spends = first(a, all_spends)?;
+
+    // at this point all_spends is a list of:
+    // (parent-coin-id puzzle-reveal amount solution . extra)
+    // where extra may be nil, or additional extension data
+
+    let mut ret = SpendBundleConditions::default();
+    let mut state = ParseState::default();
+
+    while let Some((spend, rest)) = a.next(all_spends) {
+        all_spends = rest;
+        // process the spend
+        let [parent_id, puzzle, amount, solution, _spend_level_extra] =
+            extract_n::<5>(a, spend, ErrorCode::InvalidCondition)?;
+
+        let Reduction(clvm_cost, conditions) =
+            run_program(a, &dialect, puzzle, solution, cost_left)?;
+
+        subtract_cost(a, &mut cost_left, clvm_cost)?;
+
+        let buf = tree_hash(a, puzzle);
```

Comment:
We're not caching tree hashes here, but I suppose you're probably aware. We could may add a cache parameter to `tree_hash` (it could be a linear `Vec` using the same trick we use for allocator indices). But I suppose that's intentionally not a part of this patch.

Reply:
as far as I can tell, we only compute these hashes once. I can't envision a way to accelerate these tree hashes. (unless you're talking about a future where the hard fork has activated, but that's for later)